[aotd] Fix freezing API for subclasses

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136265

Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

IvanKobzarev committed:
[aotd] Fix freezing API for subclasses

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fix freezing API for subclasses"


Original issue: 
https://github.com/pytorch/ao/issues/890

WIP - Add tests inside pytorch to repro and test a fix.

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -977,7 +977,9 @@ def aot_module_simplified(
-        tracing_context.params_flat = params_flat
+        tracing_context.params_flat = unwrap_tensor_subclasses(
+            params_flat, is_joint_structure=False
+        )
```
This does seem directionally right. also cc @mlazos

bdhirsh reviewed (approved):
sgtm!

IvanKobzarev committed:
Update on "[aotd] Fix freezing API for subclasses"


Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

[ghstack-poisoned]

Valentine233 commented:
When applying to model zoo, we also find the same error of `IndexError: list index out of range` as CI.

Is it reasonable to do the following modification?
From `max_offset_idx = max(0, len(params_flat_unwrap) - 1)` to `max_offset_idx = max(0, len(unwrapped_args_offsets) - 1)`.

IvanKobzarev committed:
Update on "[aotd] Fix freezing API for subclasses"


Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fix freezing API for subclasses"


Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[aotd] Fix freezing API for subclasses"


Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot  merge

IvanKobzarev commented:
@pytorchbot merge -f "No need to wait for torchbench runs"

atalman commented:
@pytorchmergebot revert -c ghfirst -m "Breaks internal CI sorry, need to revert"

IvanKobzarev committed:
Update on "[aotd] Fix freezing API for subclasses"


Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge
